autoprecompiles/src/powdr.rs

```diff
@@ -0,0 +1,331 @@
+use std::collections::{BTreeMap, BTreeSet};
+use std::ops::ControlFlow;
+
+use powdr_ast::analyzed::{
+    AlgebraicBinaryOperation, AlgebraicBinaryOperator, AlgebraicExpression, AlgebraicReference,
+    BusInteractionIdentity, PolyID, PolynomialType,
+};
+use powdr_ast::parsed::asm::SymbolPath;
+use powdr_ast::parsed::{
+    visitor::{ExpressionVisitable, VisitOrder},
+    NamespacedPolynomialReference, UnaryOperator,
+};
+use powdr_number::FieldElement;
+
+use crate::{BusInteractionKind, SymbolicBusInteraction};
+
+type Expression = powdr_ast::asm_analysis::Expression<NamespacedPolynomialReference>;
+
+// After powdr and lib are adjusted, this function can be renamed and the old substitute removed
+pub fn substitute_algebraic<T: Clone>(
+    expr: &mut AlgebraicExpression<T>,
+    sub: &BTreeMap<String, AlgebraicExpression<T>>,
+) {
+    expr.visit_expressions_mut(
+        &mut |expr| {
+            if let AlgebraicExpression::Reference(AlgebraicReference { name, .. }) = expr {
+                if let Some(sub_expr) = sub.get(name) {
+                    *expr = sub_expr.clone();
+                }
+            }
+            ControlFlow::Continue::<()>(())
+        },
+        VisitOrder::Pre,
+    );
+}
+
+pub fn make_refs_zero<T: FieldElement>(expr: &mut AlgebraicExpression<T>) {
+    let zero = AlgebraicExpression::Number(T::zero());
+    expr.visit_expressions_mut(
+        &mut |expr| {
+            if let AlgebraicExpression::Reference(AlgebraicReference { .. }) = expr {
+                *expr = zero.clone();
+            }
+            ControlFlow::Continue::<()>(())
+        },
+        VisitOrder::Pre,
+    );
+}
+
+pub fn is_zero<T: FieldElement>(expr: &AlgebraicExpression<T>) -> bool {
+    match expr {
+        AlgebraicExpression::Number(n) => *n == T::zero(),
+        _ => false,
+    }
+}
+
+pub fn find_byte_decomp<T: FieldElement>(
```

Comment:
The body seems to implement pattern matching, maybe there's a better name a comment for this method